docs/adr/0005-arc-pointers.md

```diff
@@ -0,0 +1,105 @@
+# Use `Arc<>` pointers instead of `HandleMap`s
+
+* Status: proposed
+* Deciders: mhammond; proposed: rfkelly, travis, jhugman, dmose
+* Date: 2021-04-19
+
+Discussion and approval: [PR 430](https://github.com/mozilla/uniffi-rs/pull/430)
+Technical Story: [Issue 419](https://github.com/mozilla/uniffi-rs/issues/419)
+
+## Context and Problem Statement
+
+uniffi currently leans heavily on the HandleMap struct in the ffi-support crate.
+This means that external consumers of uniffi-wrapped interfaces never see
+any pointers to structs - instead, they get what is (roughly) an index into
+an array, with the pointer being in that array.
+
+This has a number of safety characteristics which is particularly important for
+hand-written FFI interfaces, but does cause some issues in evolving uniffi in
+directions we consider important. In particular, it means that the lifetimes of
+uniffi wrapped structs is very rigid and thus only suitable as the `self` for
+the struct and not as generic arguments and dictionary members.
+
+This ADR considers ways to evolve the lifetime restrictions so that
+references to structs can be used ore widely than currently allowed.
+
+## Decision Drivers
+
+* We desire the ability to have more flexible life-times for our interfaces so
+  they can be stored in dictionaries or other interfaces, and be returned by
+  functions other than constructors.
+
+* We would like to keep the uniffi implementation as simple as possible while
+  providing a suitable degree of safety - in particular, a promise that it
+  should be impossible to misuse the generated bindings in a way that causes
+  us into Rust's "undefined behavior" (and in particular, avoiding things like
```

Comment:
awkward phrase "causes us into" here, do you mean "to fall into" or similar?

Reply:
thanks for calling out my hand-waving :) I mean something like "generated bindings will be kinda impossible to misuse in a way that causes bad things, like use-after-free" but not sure how to phrase that in a way that acknowledges the reality of the `unsafe` stuff we generate - I'll have another poke 👍

Reply:
I had a poke at this before I pulled your PR and we came very close to the same thing - I left a few additional words in, but happy to take them back out if you think it's getting too verbose.